File: src/main/java/org/modelingvalue/nelumbo/Or.java
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
// (C) Copyright 2018-2025 Modeling Value Group B.V. (http://modelingvalue.org)                                        ~
//                                                                                                                     ~
// Licensed under the GNU Lesser General Public License v3.0 (the 'License'). You may not use this file except in      ~
// compliance with the License. You may obtain a copy of the License at: https://choosealicense.com/licenses/lgpl-3.0  ~
// Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on ~
// an 'AS IS' BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the  ~
// specific language governing permissions and limitations under the License.                                          ~
//                                                                                                                     ~
// Maintainers:                                                                                                        ~
//     Wim Bast, Tom Brus                                                                                              ~
//                                                                                                                     ~
// Contributors:                                                                                                       ~
//     Victor Lap                                                                                                      ~
//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

package org.modelingvalue.nelumbo;

import java.io.Serial;

import org.modelingvalue.collections.List;
import org.modelingvalue.nelumbo.patterns.Functor;

public final class Or extends BinaryPredicate {
    @Serial
    private static final long serialVersionUID = -1732549494864415986L;

    private static Functor    FUNCTOR;

    static {
        KnowledgeBase.registerFunctorSetter(Or.class, f -> FUNCTOR = f);
    }

    public Or(Functor functor, List<AstElement> elements, Object[] args) {
        super(functor, elements, args[0], args[1]);
    }

    private Or(Object[] args, Or declaration) {
        super(args, declaration);
    }

    public static Or of(Node predicate1, Node predicate2) {
        return new Or(FUNCTOR, List.of(), new Object[]{predicate1, predicate2});
    }

    @Override
    public Or declaration() {
        return (Or) super.declaration();
    }

    @Override
    protected Or struct(Object[] array, Predicate declaration) {
        return new Or(array, (Or) declaration);
    }

    @Override
    public Or set(int i, Object... a) {
        return (Or) super.set(i, a);
    }

    @Override
    protected boolean isTrue(InferResult predResult) {
        return predResult.isTrueCC();
    }

    @Override
    protected boolean isFalse(InferResult predResult) {
        return false;
    }

    @Override
    protected boolean isTrue(InferResult[] predResult) {
        return false;
    }

    @Override
    protected boolean isFalse(InferResult[] predResult) {
        return predResult[0].isFalseCC() && predResult[1].isFalseCC();
    }

    @Override
    protected boolean isLeft(InferResult[] predResult) {
        return predResult[1].isFalseCC();
    }

    @Override
    protected boolean isRight(InferResult[] predResult) {
        return predResult[0].isFalseCC();
    }

    @Override
    protected InferResult add(InferResult[] predResult) {
        return predResult[0].addOr(predResult[1]);
    }

    @Override
    @SuppressWarnings("rawtypes")
    public boolean contains(Predicate cond) {
        return super.contains(cond) || predicate1().contains(cond) || predicate2().contains(cond);
    }
}

File: web-ts/src/logic/Or.ts
/**
 * Or - logical OR predicate.
 * @JAVA_REF org.modelingvalue.nelumbo.logic.Or
 */

import { List } from 'immutable';
import type { AstElement } from '../AstElement';
import { Node } from '../Node';
import type { Functor } from '../patterns/Functor';
import { BinaryPredicate } from './BinaryPredicate';
import { Predicate } from './Predicate';
import { InferResult } from '../InferResult';

/**
 * Or - logical disjunction.
 * @JAVA_REF org.modelingvalue.nelumbo.logic.Or
 */
export class Or extends BinaryPredicate {
  // @JAVA_REF org.modelingvalue.nelumbo.logic.Or#FUNCTOR
  static FUNCTOR: Functor | null = null;

  constructor(functor: Functor, elements: List<AstElement>, left: Node, right: Node) {
    super(functor, elements, left, right);
  }

  // @JAVA_REF org.modelingvalue.nelumbo.logic.Or#of(Predicate, Predicate)
  static of(predicate1: Predicate, predicate2: Predicate): Or {
    return new Or(Or.FUNCTOR!, List<AstElement>(), predicate1, predicate2);
  }

  protected static fromDataOr(data: unknown[], declaration?: Node): Or {
    const or = Object.create(Or.prototype) as Or;
    (or as unknown as { _data: unknown[] })._data = data;
    (or as unknown as { _declaration: Node })._declaration = declaration ?? or;
    (or as any)._binding = null;
    (or as any)._hashCodeCached = false;
    (or as any)._hashCode = 0;
    (or as any)._nrOfUnbound = -1;
    return or;
  }

  protected override struct(data: unknown[], declaration?: Node): Or {
    return Or.fromDataOr(data, declaration ?? this.declaration());
  }

  override declaration(): Or {
    return super.declaration() as Or;
  }

  override set(i: number, ...a: unknown[]): Or {
    return super.set(i, ...a) as Or;
  }

  // @JAVA_REF org.modelingvalue.nelumbo.logic.Or#isTrue(InferResult, int)
  protected isTrueSingle(predResult: InferResult, _i: number): boolean {
    return predResult.isTrueCC();
  }

  // @JAVA_REF org.modelingvalue.nelumbo.logic.Or#isFalse(InferResult, int)
  protected isFalseSingle(_predResult: InferResult, _i: number): boolean {
    return false;
  }

  // @JAVA_REF org.modelingvalue.nelumbo.logic.Or#isUnknown(InferResult, int)
  protected isUnknownSingle(_predResult: InferResult, _i: number): boolean {
    return false;
  }

  // @JAVA_REF org.modelingvalue.nelumbo.logic.Or#isTrue(InferResult[])
  protected isTrueBoth(_predResult: InferResult[]): boolean {
    return false;
  }

  // @JAVA_REF org.modelingvalue.nelumbo.logic.Or#isFalse(InferResult[])
  protected isFalseBoth(predResult: InferResult[]): boolean {
    return predResult[0].isFalseCC() && predResult[1].isFalseCC();
  }

  // @JAVA_REF org.modelingvalue.nelumbo.logic.Or#isLeft(InferResult[])
  protected isLeft(predResult: InferResult[]): boolean {
    return predResult[1].isFalseCC();
  }

  // @JAVA_REF org.modelingvalue.nelumbo.logic.Or#isRight(InferResult[])
  protected isRight(predResult: InferResult[]): boolean {
    return predResult[0].isFalseCC();
  }
}
